packages/java-parser/src/productions/expressions.js

```diff
@@ -561,33 +561,49 @@ function defineRules($, t) {
     ]);
   });
 
+  // https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-Pattern
   $.RULE("pattern", () => {
-    $.SUBRULE($.primaryPattern);
-    $.OPTION(() => {
-      $.CONSUME(t.AndAnd);
-      $.SUBRULE($.binaryExpression);
-    });
-  });
-
-  $.RULE("primaryPattern", () => {
     $.OR([
       {
-        ALT: () => {
-          $.CONSUME(t.LBrace);
-          $.SUBRULE($.pattern);
-          $.CONSUME(t.RBrace);
-        }
+        GATE: () => this.BACKTRACK_LOOKAHEAD($.recordPattern),
```

Comment:
I think we could switch the two options: as there is recursion in the pattern/recordPattern definition, I would suggest backtracking on the typePattern instead

Reply:
Good point! Done.